Задание 6 спринта

polyakovanan opened the pull request:
Реализовал связку "двунаправленного списка" и хэшмапы для хранения истории обращения к задачам с учетом исключения повторов. Добавил возможность удалять из истории записи.

Реализовал интерфейс Clonable во всех классах задач, чтобы защитить данные от изменений извне. Теперь в менеджер попадают клоны входящих задач и отдаются так же клоны.

Покрыл тестами новый механизм хранения истории и защиты данных.

Реализовал пользовательский сценарий в Main

polyakovanan committed:
Реализовал связку "двунаправленного списка" и хэшмапы для хранения истории обращения к задачам с учетом исключения повторов. Добавил возможность удалять из истории записи.

Реализовал интерфейс Clonable во всех классах задач, чтобы защитить данные от изменений извне. Теперь в менеджер попадают клоны входящих задач и отдаются так же клоны.

Покрыл тестами новый механизм хранения истории и защиты данных.

Реализовал пользовательский сценарий в Main

polyakovanan committed:
Merge branch 'main' into sprint_6-solution

polyakovanan committed:
Правки по 6 спринту, итерация 1.
Изменения в код стайле. Нормальный такой рофл, конечно. Хоть бы сразу весь список выдавало, но не по одной строке

polyakovanan committed:
Правки по 6 спринту, итерация 2.
Методы должны быть отделены пустыми строками, ага...

polyakovanan committed:
Правки по 6 спринту, итерация 3.
В чем проблема весь список замечаний по код стайлу выдать?

polyakovanan committed:
Правки по 6 спринту, итерация 4.
У меня слов просто нет, ребята

polyakovanan committed:
Правки по 6 спринту, итерация 5.
...

polyakovanan committed:
Правки по 6 спринту, итерация 6.
...

polyakovanan committed:
Правки по 6 спринту, итерация 7.
...

TyutterinYakov commented on src/ru/terralink/kanban/model/Epic.java:
```diff
@@ -69,10 +70,20 @@ private void calcStatus(){
     }
 
+    @Override
+    public Object clone() {
```
Отлично, молодец!

polyakovanan replied:
Спасибо

TyutterinYakov commented on src/ru/terralink/kanban/service/InMemoryHistoryManager.java:
```diff
@@ -1,31 +1,92 @@
-        this.maxSize = size;
-        taskHistory = new LinkedList<>();
+    private final PatheticLinkedList<Task> taskHistory;
+    private final Map<Integer,Node> taskHash;
```
Так как класс `Node` является параметризированным, хорошо было бы указать тип в даймонд операторе)

polyakovanan replied:
Добавил тип везде, где можно. Теперь InMemoryHistoryManager работает только с Node<Task>

TyutterinYakov commented on src/ru/terralink/kanban/service/InMemoryHistoryManager.java:
```diff
@@ -1,31 +1,92 @@
+    }
+
+    private void removeNode(Node node) {
+        Node nextNode = node.getNext();
```
Можно было бы здесь добавить проверку значению `node` на `null`, чтобы не дублировать ее в других местах)

polyakovanan replied:
Занес проверку на null внутрь метода

TyutterinYakov commented on src/ru/terralink/kanban/service/InMemoryHistoryManager.java:
```diff
@@ -1,31 +1,92 @@
+        if (prevNode != null)  {
+            prevNode.setNext(nextNode);
+            node.setPrev(null);
+        }
```
Можно было бы записать логику из условия со строк `56-58` в блоке `else` здесь) Так как, если перед узлом нет другого - он голова)

polyakovanan replied:
Совместил логику

TyutterinYakov commented on src/ru/terralink/kanban/service/InMemoryHistoryManager.java:
```diff
@@ -1,31 +1,92 @@
+        if (nextNode != null) {
+            nextNode.setPrev(prevNode);
+            node.setNext(null);
+        }
```
Аналогично, но логику со строк 59-61

polyakovanan replied:
Совместил логику

TyutterinYakov commented on src/ru/terralink/kanban/service/InMemoryHistoryManager.java:
```diff
@@ -1,31 +1,92 @@
+        private Node<E> last;
+
+        public Node<E> linkLast(E element) {
+            Node<E> lastNode = this.last;
```
Можно было бы записать без ввода переменной, перенеся установку значения для `last` - после условия ниже)

polyakovanan replied:
Вынес обновление ссылки в last после всех манипуляций

TyutterinYakov commented on src/ru/terralink/kanban/service/InMemoryHistoryManager.java:
```diff
@@ -1,31 +1,92 @@
+            List<E> items = new ArrayList<>();
+            Node currentNode = first;
+            while (currentNode != null) {
+                items.add((E) currentNode.getItem());
```
Приведение более не потребуется)

polyakovanan replied:
Вынес тип E в параметризацию Node<E>

TyutterinYakov commented on src/ru/terralink/kanban/service/InMemoryHistoryManager.java:
```diff
@@ -1,31 +1,92 @@
+
+    @Override
+    public void remove(int id) {
+        Node node = taskHash.get(id);
```
Теперь можно будет записать в одну строку:
`removeNode(taskHash.remove(id));`

polyakovanan replied:
Записал одной строкой

TyutterinYakov commented on src/ru/terralink/kanban/service/InMemoryHistoryManager.java:
```diff
@@ -1,31 +1,92 @@
-            taskHistory.remove(0);
+        int taskId = task.getId();
+        Node node = taskHash.get(taskId);
+        if (node != null) {
```
Теперь можно будет записать в одну строку:
`removeNode(taskHash.get(id));`

polyakovanan replied:
Записал одной строкой

TyutterinYakov reviewed (commented):
Добрый день! В целом - хорошо) Отметил то, что следует исправить и то, что можно сделать лучше) Хорошего дня! Если возникнут вопросы по ревью - пиши в пачку, буду рад помочь🌅

polyakovanan committed:
Правки по 6 спринту, итерация 2.1
Добавил типизацию Node везде, где нужно. InMemoryHistoryManager теперь везде работает только с нодами типа Task.
Поправил PatheticLinkedList.getItems(). Node<E> вместо приведения.

Проверка на node == null внутри removeNode, а не перед вызовом. -> Вызов removeNode в одну строку. + уплотнил логику перепривязки нод.

polyakovanan committed:
Правки по 6 спринту, итерация 2.1
Избавился от промежуточной переменной в linkLast, перенеся присвоение новой ноды в last после всех остальных манипуляций

TyutterinYakov commented on src/ru/terralink/kanban/service/InMemoryHistoryManager.java:
```diff
@@ -1,31 +1,85 @@
+        if (prevNode != null)  {
+            prevNode.setNext(nextNode);
+            node.setPrev(null);
+        } else if (node.equals(taskHistory.first)) {
```
Можно просто `else`, так как это условие, по нашей заложенной логике, полностью противоположно тому, что в блоке `if`

TyutterinYakov commented on src/ru/terralink/kanban/service/InMemoryHistoryManager.java:
```diff
@@ -1,31 +1,85 @@
+        if (nextNode != null) {
+            nextNode.setPrev(prevNode);
+            node.setNext(null);
+        } else if (node.equals(taskHistory.last)) {
```
Можно просто `else`, так как это условие, по нашей заложенной логике, полностью противоположно тому, что в блоке `if`

TyutterinYakov commented on src/ru/terralink/kanban/service/InMemoryHistoryManager.java:
```diff
@@ -1,31 +1,85 @@
+    @Override
+    public void remove(int id) {
+        removeNode(taskHash.remove(id));
+        taskHash.remove(id);
```
Выше мы уже произвели удаление из мапы)

TyutterinYakov commented on src/ru/terralink/kanban/service/InMemoryHistoryManager.java:
```diff
@@ -1,31 +1,85 @@
-        }
-        taskHistory.add(task);
+        int taskId = task.getId();
+        removeNode(taskHash.remove(taskId));
```
Просто для справки: Удаление из мапы необязательно, в данном случае, так как при вызове метода `put` ниже - значение уже будет заменено)

TyutterinYakov reviewed (approved):
Добрый вечер! Отметил оставшиеся небольшие недочеты, но думаю, что ты справишься с ними вне ревью) Успехов в обучении и прекрасного вечера!🌃🎇

polyakovanan committed:
Правки по 6 спринту, итерация 2.2
Подрихтовал removeNode() и убрал избыточные удаления из taskHash
